### PR TITLE
feat: Add UseGuards detection to controller methods

### PR DESCRIPTION
**Pull request by Fishbowler**
## Description

Adds a lint to ensure that any endpoint method has a `@UseGuards` decorator, either at the method or class level.

## Commentary

It's super useful for ensuring correct endpoints have appropriate authentication + authorisation, but it's useless for an open API. 

## Outstanding work

I've added it to Recommended for testing, but I don't think it really belongs there. Could go into Recommended in an "off" state, allowing people to override it?

It doesn't yet have docs - I'll add those as soon as I know where it's going (i.e. that this will be accepted as a PR in some state)

### CODE REVIEW

File: src/rules/apiMethodsShouldBeGuarded/apiMethodsShouldBeGuarded.test.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+                @ApiOperation({ summary: 'Check if registration is available by Company name' })
+                async canRegister(
+                    @Body() { companyName }: CompanyCanRegisterDto,
+                    @Query('countryCode') countryCode?: string,
+                ): Promise<TResultResponse<boolean>> {
+                    const result = await this.companiesService.canRegisterName(companyName, countryCode)
+
+                    return { result }
+                }
+            }`,
```
Comment (Fishbowler) [inline src/rules/apiMethodsShouldBeGuarded/apiMethodsShouldBeGuarded.test.ts:93]: I'm not a big fan of this complex example, but it felt like it changed enough aspects (Post path, async inside and outside, annotations on method params) to be useful. Happy to bin it if it's too ugly.

**Comment by Fishbowler** [comment]
I've updated the setting in Recommended to be Off, which gets it into the import, but I think makes this safe to include for everyone.

**Comment by darraghoriordan** [comment]
Hey, thanks for the PR!

I wonder if this scenario is better handled with a global guard instead. You can attach the guard directly to the app and nestjs will ensure all routes are protected

```
app.useGlobalGuards(new RolesGuard());
```

(assuming RolesGuard is some custom guard you created)

then for the occasional public route, instead of `@Guard` you could instead create a `@PublicEndpoint()` metadata decorator. Then you could update `RolesGuard` to check for the presence of this metadata and allow access to anyone if some IsPublic metadata is present.

**Comment by Fishbowler** [comment]
Nest provides for separate Authentication and Authorization, where the latter is implemented in Guards, and can take advantage of other controller/endpoint metadata (like a RolesGuard using a Role annotation, from your example).

We hit a problem where with different _authentication_ mechanisms (e.g. Firebase, Google SSO, internal) conflating the authorization to work across them was messy, could compromise the SRP, and so add risk to the critical nature of the component. Further, splitting guards out to be explicit at a controller or endpoint level ensures that it's a conscious developer choice in our context.

Definitely a rule that wants to be `off` by default, since it won't apply to all situations

I wonder if there's a way to detect the useGlobalGuards, to make it even less intrusive?

**Comment by darraghoriordan** [comment]
> Nest provides for separate Authentication and Authorization, where the latter is implemented in Guards, and can take advantage of other controller/endpoint metadata (like a RolesGuard using a Role annotation, from your example).
> 
> We hit a problem where with different _authentication_ mechanisms (e.g. Firebase, Google SSO, internal) conflating the authorization to work across them was messy, could compromise the SRP, and so add risk to the critical nature of the component. Further, splitting guards out to be explicit at a controller or endpoint level ensures that it's a conscious developer choice in our context.
> 
> Definitely a rule that wants to be `off` by default, since it won't apply to all situations
> 
> I wonder if there's a way to detect the useGlobalGuards, to make it even less intrusive?

ah fair enough. sounds like a tricky situation. i'll take a closer look at this over the next couple of days then hopefully and get it merged

**Review by darraghoriordan — APPROVED** [review]

**Comment by darraghoriordan** [comment]
Hey it all lgtm. it's turned off by default anyway so should be fine.

Any chance you can add a section to README.md describing what the rule does with one or two examples?

**Comment by Fishbowler** [comment]
Done 👍

**Pull request merged**